Milestone 1 of Internal Process-level Fault Tolerance (RFC #27866)

fangyuchu opened the pull request:
**Co-authored with:** @zWaNg3 @a798347923 @205150940
## Purpose

### Summary

This PR implements **Milestone 1** of RFC #27866, introducing a fault tolerance framework for vLLM.
It provides mechanisms to **detect runtime exceptions**, **pause execution**, **report fault events**, and allow upper orchestration layers to **apply fault tolerance methods** (initially supporting retry).

**（Updates 18/2/2026）**
Since the complete fault tolerance framework introduces substantial code changes, we also plan to split the original PR into three smaller and more reviewable parts: fault reporting, pause-on-error, and fault handling interface.
The detailed split plan can be found [here](https://docs.google.com/document/d/1o0uAP_MpjyU4aN0CEZ6PUgt88KLap1jccQKlJo4-TE4/edit?usp=sharing).
Links to the split PRs:
- Fault reporting: https://github.com/vllm-project/vllm/pull/34833
- Pause-on-error: https://github.com/vllm-project/vllm/pull/38534
- Fault handling interface: https://github.com/vllm-project/vllm/pull/40468

### Supported Functionality

- Detect exceptions raised inside a vLLM instance and pause its execution.
- Publish fault events and expose rank/EngineCore status to the upper orchestration framework.
- Provide a retry API for transient and recoverable faults (e.g., network jitter).

### Usage

Enable the Fault Tolerance feature via configuration, for example:

```bash
vllm serve \
  --model xxxxx \
  --data-parallel-size 8 \
  --enable-expert-parallel \
  --tensor-parallel-size 1 \
  --enforce-eager \
  --enable-fault-tolerance \
  --engine-recovery-timeout 120 \
  --external-fault-notify-port 23456
```
- When an exception is raised in any rank (tested via code injection or killing a worker process), vLLM reports the event via a ZMQ PUB socket.

  - Example: Subscribe to port 23456 on topic vllm_fault to receive events.
- The status of each EngineCore can be queried via API:
```bash
GET /fault_tolerance/status
```
- Retry can be triggered via API:
```css
POST /fault_tolerance/apply
Body: {"fault_tolerance_instruction": "retry", "fault_tolerance_timeout": 30}
```

### Implementation Notes
Overall architecture:
![Framework of vLLM Internal Fault Tolerance](https://github.com/user-attachments/assets/bb673db0-e923-422e-88ee-409bab455213)

```
**ClientSentinel (for MPClient)**
  ├─ Receives:
  │    - Exception / failure reports from EngineCoreSentinel
  │    - Fault-tolerance commands from the API Server
  ├─ Provides:
  │    - Query & subscription interface for upper-level frameworks
  └─ Forwards:
       - Fault-tolerance commands to EngineCoreSentinel
    ↓ ZMQ ROUTER/DEALER
**EngineCoreSentinel (per DP rank)**
  ├─ Monitors:
  │    - EngineCore busy_loop via a shared fault_signal_q
  ├─ Controls:
  │    - busy_loop_active flag
  └─ Commands:
       - WorkerSentinel via ZMQ
    ↓ ZMQ ROUTER/DEALER
**WorkerSentinel (per worker)**
  ├─ Monitors:
  │    - N/A (passive, command-driven)
  ├─ Executes received commands:
  │    - pause_by_signal
  │    - pause_by_abort_communicators
  |    -  ...
  └─ Reports:
       - Success / failure back to EngineCoreSentinel
```

- We actually implemented two approaches to pause healthy ranks. Due to the complexity of collective communications, we currently adopt the hard pause method as the default.
  - Soft pause: set pause flag variables for EngineCore/Worker to check.
  - Hard pause (default): set flags and abort process groups/communicators.

- NCCL: All communicators are set as nonblocking when fault tolerance is enabled, allowing safe use of [ncclCommAbort](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/comms.html#c.ncclCommAbort) as suggested by [NCCL](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/communicators.html#fault-tolerance).

- Currently, when querying EngineCore status via the API: EngineCores paused using the signal flag are reported as healthy. EngineCores paused due to aborted communicators are reported as unhealthy. Further refinements to the status reporting will be implemented in follow-ups.

We believe that **this implementation is compatible with RFC #27774 and #27908** to isolate fault ranks through scale_down. This implementation  establishes the overall fault tolerance framework within vLLM, providing mechanisms for detecting, pausing, reporting, and performing fault tolerance. Within this framework, **fault isolation through elastic expert parallelism (#20323)** is treated as **one of the fault tolerance methods**. As a result, this implementation and the elastic expert parallelism proposal are designed to be compatible.

We welcome further discussion and feedback from the community to refine and extend these fault tolerance mechanisms.


## Test Plan

- Tested with DT
  - pytest tests/v1/engine/test_client_guard.py
  - pytest tests/v1/engine/test_engine_core_guard.py
- End-to-end tests.
  - End-to-end tests conducted on Qwen3-30B-A3B with configurations: DP4/TP2 and DP8/TP1.
  - Two fault injection types:
    - Manually raising exceptions at different code locations.
    - Killing certain worker processes.
  - Verified:
    - Pause behavior
    - Fault reporting
    - Retry functionality (for code injections)

## Test Result
**DT results**

- pytest tests/v1/engine/test_client_guard.py
  - tests/v1/engine/test_client_guard.py::test_client_guard_initialization PASSED
  - tests/v1/engine/test_client_guard.py::test_handle_fault PASSED
  - tests/v1/engine/test_client_guard.py::test_fault_receiver PASSED
  - tests/v1/engine/test_client_guard.py::test_fault_receiver_unhealthy PASSED
  - tests/v1/engine/test_client_guard.py::test_shutdown_guard PASSED
  - tests/v1/engine/test_client_guard.py::test_handle_fault_async PASSED
- pytest tests/v1/engine/test_engine_core_guard.py
  - tests/v1/engine/test_engine_core_guard.py::test_engine_core_guard_initialization PASSED
  - tests/v1/engine/test_engine_core_guard.py::test_run_handle_instruction[pause] PASSED
  - tests/v1/engine/test_engine_core_guard.py::test_run_handle_instruction[retry] PASSED


**End-to-end Test Results**
We validated both the **functional correctness** and **model precision** after applying the retry-based fault tolerance mechanism. All fault tolerance features behaved as expected — including fault detection, pause/resume, event reporting, and retry recovery.

The precision tests showed no regression, indicating that the retry mechanism does not affect the model’s output quality.

Here are some vLLM log demonstrations from the end-to-end tests:
**1. Simulated Exception**
[DP4_TP2_Exception_Retry.log](https://github.com/user-attachments/files/23495720/DP4_TP2_Exception_Retry.log)
A fault was manually triggered by raising a `RuntimeError("Test Error")` inside `gpu_model_runner.execute_model` on DP rank0.
- The instance started at **23:02:12** and finished the initialization at **23:02:44**, taking roughly **32 seconds** to startup, and serving requests afterwards.
- At **23:03:02**, the exception was raised and propagated to the client. The ClientGuard immediately began executing the Pause operation (_"[FaultHandler] Pause operation is best-effort ..."_).
- By **23:03:03**, all EngineCore components had stopped computation and entered the paused state (_"[EngineCoreGuard_x] Command (pause) succeeded: True"_), and the runtime status could be queried via GET /fault_tolerance/status.
- At **23:03:14**, an external Retry command was issued, and by **23:03:16**, the instance successfully resumed service (_"[EngineCoreGuard_x] Command (retry) succeeded: True"_)

**The recovery process took only 2 seconds, compared with 32 seconds for a full restart recovery**, demonstrating significantly improved fault recovery efficiency. After retry, normal inference resumed (_"Engine 000: Avg prompt throughput: 3.9 tokens/s, Avg generation throughput: 4.3 tokens/s, ..."_).

We also tested with ray backend with DP8/TP1, and here is the log:
[Ray_DP8_TP1_Exception_Retry.log](https://github.com/user-attachments/files/23496306/Ray_DP8_TP1_Exception_Retry.log)

**2. Random process kill simulation**
[DP8_TP1_KillProcess.log](https://github.com/user-attachments/files/23496205/DP8_TP1_KillProcess.log)
In another test, we simulated faults by randomly killing one of the worker processes (23:23:49, EngineCore DP4). 
After the kill, the instance automatically entered the paused state. 
As no fault-tolerance command was received within the pre-defined timeout (engine-recovery-timeout, 120s), the instance exited gracefully(_"23:25:49 [core.py:389] [BusyLoopWrapper] Fault tolerance instruction not received within timeout. Proceeding with default exception handling."_).


---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

fangyuchu commented on vllm/v1/engine/core.py:
```diff
@@ -73,6 +86,318 @@
+        )
+        for identity in identities:
+            response = worker_responses.get(identity)
+            if response is None or not response.get("success", False):
```
In the `wait_for_instruction_result` function, the line `target_identities = set(target_identities)` creates a new set as a shallow copy of the input target_identities. This means any modifications made to target_identities within `wait_for_instruction_result` will only affect this local copy, not the original identities set passed from `_execute_worker_method`.

ruisearch42 reviewed (commented):
Thanks for the PR!

First pass. Just some nitpicks.

I think it makes sense overall. Any end-to-end tests/demonstration that this is functional and actually helps recover from faults (considering there are quite some changes added)?

fangyuchu commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -60,6 +74,183 @@
+                nccl_comm.available = active
+                nccl_comm.disabled = not active
+
+    def restart_worker(self):
```
The name has been changed to restore_worker.

fangyuchu commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -60,6 +74,183 @@
+        ).start()
+
+    def _make_worker_logger(self):
+        prefix = f"[WorkerGuard_dp{self.dp_rank}_tp{self.tp_rank}_pp{self.pp_rank}] "
```
The order is corrected.

hmellor commented on vllm/config/vllm.py:
```diff
@@ -10,7 +10,7 @@
 import time
 from contextlib import contextmanager
-from dataclasses import replace
+from dataclasses import field, replace
```
```suggestion
from dataclasses import replace
```

hmellor commented on vllm/config/vllm.py:
```diff
@@ -107,6 +108,10 @@ class VllmConfig:
     """The configurations for event publishing."""
     ec_transfer_config: ECTransferConfig | None = None
     """The configurations for distributed EC cache transfer."""
+    fault_tolerance_config: FaultToleranceConfig = field(
```
```suggestion
    fault_tolerance_config: FaultToleranceConfig = Field(
```

fangyuchu replied:
Thanks for the review. Fixed as advised.

hmellor reviewed (commented):
Just a couple of nits on the config code, not a full review

tzulingk commented on vllm/distributed/parallel_state.py:
```diff
@@ -1040,21 +1060,29 @@ def get_inner_dp_world_group() -> GroupCoordinator:
+    ranks: list[int],
+    local_rank: int,
+    backend: str,
+    enable_fault_tolerance: bool = False,
```
nit: consider using FaultToleranceConfig as input param?

tzulingk commented on vllm/v1/engine/utils.py:
```diff
@@ -154,12 +196,53 @@ def __init__(
         """Shutdown all procs."""
         self._finalizer()
 
+    def start_engine_core_monitor(self):
```
this seems to be very similar to start_engine_core_monitor in vllm/v1/engine/core_client.py. Can we consolidate them?
ex
```
def report_engine_death(
    died_proc,
    sentinel,
    sentinels_list,
    engine_down_socket,
    logger
):
    """
    Report engine death via fault info and clean up.
    
    Args:
        died_proc: The dead process
        sentinel: The sentinel that triggered
        sentinels_list: List to remove sentinel from
        engine_down_socket: ZMQ socket to report on
        logger: Logger instance
    """
    # Extract rank from process name
    match = re.match(r"EngineCore_DP(\d+)", died_proc.name)
    engine_rank = match.group(1) if match else "unknown"
    
    # Create fault info
    fault_info = FaultInfo(
        type="engine_core dead",
        message=f"Engine core proc {died_proc.pid} ({died_proc.name}) died unexpectedly.",
        engine_id=engine_rank,
        additional_info=None,
    )
    
    # Report via ZMQ
    engine_down_socket.send_multipart([b"", fault_info.serialize().encode("utf-8")])
    
    # Cleanup
    if isinstance(sentinel, int) and sentinel in sentinels_list:
        sentinels_list.remove(sentinel)
    
    # Log
    logger.error("Engine core proc %s died unexpectedly", died_proc.name)
    
    return engine_rank, fault_info
```
And use in both files

tzulingk commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -61,6 +75,184 @@
     from vllm.model_executor.model_loader.tensorizer import TensorizerConfig
 
 
+class WorkerSentinel:
```
there appear to be multiple Sentinel types:

- EngineCoreSentinel - Monitors EngineCore
- WorkerSentinel - Monitors workers
- ClientSentinel - Monitors client (mentioned in tests)
Can we consolidate them?
ex
```
class BaseSentinel:
    """
    Base class for monitoring and fault detection.
    Handles common patterns: command queue, fault reporting, monitoring loop.
    """
    
    def __init__(self, name, fault_report_socket, logger_func):
        # Common attributes
        self.name = name
        self.sentinel_active = True
        self.command_queue = queue.Queue()
        self.fault_queue = queue.Queue()
        self.fault_report_socket = fault_report_socket
        self.logger = logger_func
        
        # Status tracking
        self.monitored_entities_status = {}  # id -> "Running" | "Paused" | "Dead"
        
    def run(self):
        """Main monitoring loop - template method pattern."""
        self.logger(f"{self.name} starting", level="info")
        
        while self.sentinel_active:
            # 1. Check for faults (subclass-specific)
            fault = self._check_for_faults()
            if fault:
                self._handle_fault(fault)
            
            # 2. Process commands (common logic)
            self._process_pending_commands()
            
            # 3. Sleep or wait (subclass can override)
            self._wait_for_next_iteration()
    
    # Abstract methods (subclasses must implement)
    def _check_for_faults(self):
        """Subclass implements: how to detect faults."""
        raise NotImplementedError
    
    # Shared methods
    def _handle_fault(self, fault):
        """Common fault handling logic."""
        fault_info = self._create_fault_info(fault)
        self._publish_fault(fault_info)
        self.fault_queue.put(fault)
    
    def _process_pending_commands(self):
        """Process commands from external orchestrator."""
        try:
            cmd = self.command_queue.get_nowait()
            self._execute_command(cmd)
        except queue.Empty:
            pass
    
    def _execute_command(self, cmd):
        """Execute orchestrator command (pause, retry, etc.)."""
        method_name, params = deserialize_command(cmd)
        method = getattr(self, method_name, None)
        if method:
            result = method(**params)
            self._send_command_result(result)
    
    def _publish_fault(self, fault_info):
        """Publish fault via ZMQ."""
        self.fault_report_socket.send_multipart(
            [b"", fault_info.serialize()]
        )
    
    def _create_fault_info(self, fault):
        """Create FaultInfo from detected fault."""
        return FaultInfo(
            type=fault.type,
            message=str(fault.message),
            entity_id=fault.entity_id,
        )
    
    def _wait_for_next_iteration(self):
        """Default: small sleep. Subclasses can override."""
        time.sleep(0.1)
    
    def shutdown(self):
        """Common shutdown logic."""
        self.sentinel_active = False
        self.logger(f"{self.name} shutting down", level="info")
```

fangyuchu replied:
Thanks again for the great suggestion. We have now refactored the Sentinels with a common base class, BaseSentinel, so that all sentinel types share the same parent.

fangyuchu commented on vllm/distributed/parallel_state.py:
```diff
@@ -323,14 +326,31 @@ def __init__(
+            else:
+                device_group = torch.distributed.new_group(
+                    ranks, backend=torch_distributed_backend
+                )
```
Revised as suggested. Thanks!

fangyuchu commented on vllm/v1/engine/core.py:
```diff
@@ -73,6 +85,344 @@
+        for identity in identities:
+            response = worker_responses.get(identity)
+            if response is None or not response.get("success", False):
+                all_success = False
```
Revised as suggested. Thanks!

fangyuchu commented on vllm/v1/engine/utils.py:
```diff
@@ -64,6 +80,14 @@ class EngineZmqAddresses:
+    fault_report_addr: str | None = None
+    # ZMQ client_cmd socket address of client sentinel
+    client_cmd_addr: str | None = None
+    # identities of engine_core_sentinel
```
We add the following comment:  Identities of engine core DEALER sockets, keyed by engine index. These identities are used by the ClientSentinel (ROUTER) to route messages to the corresponding engine core.

hmellor commented on vllm/config/fault_tolerance.py:
```diff
@@ -0,0 +1,74 @@
+
+
+@config
+@dataclass
```
`config` is a dataclass transform now

```suggestion
@config
```

tzulingk commented on vllm/distributed/device_communicators/pynccl_wrapper.py:
```diff
@@ -360,10 +371,32 @@ def __init__(self, so_file: str | None = None):
+                    "NCCL_CHECK: ncclInProgress returned but no communicator "
+                    "provided (required for non-blocking NCCL checks)."
+                )
+            while True:
```
When I used non-blocking mode, I actually got lots of different issues when running MoE models. 
But I will trust that you have make things right, maybe I just wasn't doing thing correctly.

fangyuchu committed:
Milestone 1 of Internal Process-level Fault Tolerance

* Milestone 1 of Internal Process-level Fault Tolerance (#61)

* feat(fault-tolerance): add class skeletons for fault tolerance

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* config: add configuration options for fault tolerance

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* 增加generate_identity和generate_identitys函数 Generate a unique identity for ZMQ ROUTER node

* add service startup configuradtion fault report addr

* add init WorkerGuard

* add engine_core_cmd_addr、fault_report_addr、client_cmd_addr、engine_core_identitys in EngineZmqAddresses
init engine_core_cmd_addr、fault_report_addr、client_cmd_addr in launch_core_engines func
add _report_engine_dead func in CoreEngineProcManager

* init ClientGuard
init EngineZmqAddresses engine_core_identitys

* init EngineCoreGuard

* change generate_identitys to generate_identity_group

* code typesetting is optimized

* code typesetting is optimized

* changed code format ensure every line < 88 chars

* changed code format ensure every line < 88 chars
fix error Value of type "dict[Any, Any] | None" is not indexable  [index]

* fix bug
Error: vllm/v1/engine/utils.py:122:89: E501 Line too long (117 > 88)
Error: vllm/v1/engine/utils.py:1059:9: F402 Import `uuid` from line 6 shadowed by loop variable

* fix
Error: vllm/v1/engine/utils.py:1045: error: Need type annotation for "uuids" (hint: "uuids: set[<type>] = ...")  [var-annotated]

* fix
error: Value of type "dict[Any, Any] | None" is not indexable  [index]

* fix
error: Value of type "dict[Any, Any] | None" is not indexable  [index]

Signed-off-by: a798347923 <2645302020@qq.com>

* add _send_msg in EngineCoreGuard

Signed-off-by: a798347923 <2645302020@qq.com>

* add import torch.cuda

* add _recv_cmd function docstring that clearly explains the meaning of the return value.

* changed recv_fault_msg to recv_msg
add ClientGuard __init__ func parameter types

* add engine monitor

Signed-off-by: TianZhuo <2770730562@qq.com>

* Delete requirements/test.txt~

Signed-off-by: a798347923 <39047817+a798347923@users.noreply.github.com>

* Delete vllm/v1/engine/core_client.py~

Signed-off-by: a798347923 <39047817+a798347923@users.noreply.github.com>

* simply _send_msg and _recv_cmd in EngineCoreGuard

* simply recv_msg in ClientGuard

* engine: add fault tolerance features for EngineCore.

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* engine: add timeout mechanism in retry.

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* add engine monitor

* Delete vllm/v1/engine/exceptions.py~

Signed-off-by: 205150940 <112750056+205150940@users.noreply.github.com>

* updata actor_index

* updata enginedead flag

* handle fault and report exception

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* fix engine_actor

* fix engine_actor fault_info

* handle fault and report exception

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* delete num_identity

* changed try expect

* fix debug error

* fix one bug.

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* add fault_report_addr in FaultToleranceConfig

* add handle fault&get_fault_info api

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* remove fault_report_address in CoreEngineActorManager __init__

Signed-off-by: a798347923 <2645302020@qq.com>

* ruff format

Signed-off-by: a798347923 <2645302020@qq.com>

* add handle fault&get_fault_info api

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* fix one bug.

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* add fault_report_port in FaultToleranceConfig

Signed-off-by: a798347923 <2645302020@qq.com>

* add zmq_addr concatenate with fault_report_addr and fault_report_port

Signed-off-by: a798347923 <2645302020@qq.com>

* fault reporter bug fix

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* fault reporter bug fix

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* fault reporter bug fix

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* fault reporter bug fix

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* fault reporter bug fix

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* fault reporter bug fix

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* fix some bug

* fault reporter bug fix

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* fault reporter bug fix

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* remove fault_report_addr in FaultToleranceConfig

Signed-off-by: a798347923 <2645302020@qq.com>

* refactor: relocate method serialization functions to serial_util.py

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* fix actor bug

* fix actor bug

* add engine_core_cmd_addr in FaultToleranceConfig

Signed-off-by: a798347923 <2645302020@qq.com>

* add and use _stop_worker_execution in EngineCoreGuard

Signed-off-by: a798347923 <2645302020@qq.com>

* add and use run in WorkerGuard

Signed-off-by: a798347923 <2645302020@qq.com>

* fix actor bug

* fix bug

* fix sentinel

* fix bug vllm/v1/engine/core.py:847: error: Missing positional argument "tp_size" in call to "EngineCoreGuard"

Signed-off-by: a798347923 <2645302020@qq.com>

* fix bug error: Missing positional arguments "length", "byteorder" in call to "to_bytes" of "int"

Signed-off-by: a798347923 <2645302020@qq.com>

* fix bug in fault tolerance mode

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* fix bug in fault tolerance mode

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* change fault_report_port to internal_fault_report_port
add external_fault_notify_port

Signed-off-by: a798347923 <2645302020@qq.com>

* change fault_report_port to internal_fault_report_port
add external_fault_notify_port

Signed-off-by: a798347923 <2645302020@qq.com>

* add _recv_cmd func
use deserialize_method_call and run_method in run func

Signed-off-by: a798347923 <2645302020@qq.com>

* Update core.py

fix bug error: Need type annotation for "kwargs" (hint: "kwargs: dict[<type>, <type>] = ...")

Signed-off-by: a798347923 <39047817+a798347923@users.noreply.github.com>

* add self.ctx.term() in shutdown()

Signed-off-by: a798347923 <2645302020@qq.com>

* changed import deserialize_method_call,serialize_method_call

Signed-off-by: a798347923 <2645302020@qq.com>

* changed init worker_guard in init_device

Signed-off-by: a798347923 <2645302020@qq.com>

* Update core.py

add import serialize_method_call

Signed-off-by: a798347923 <39047817+a798347923@users.noreply.github.com>

* Update gpu_worker.py

changed init WorkerGuard in init_device

Signed-off-by: a798347923 <39047817+a798347923@users.noreply.github.com>

* Update gpu_worker.py

FIX BUG self.worker_guard: WorkerGuard|None = None

Signed-off-by: a798347923 <39047817+a798347923@users.noreply.github.com>

* Update gpu_worker.py

fix bug error: Argument 1 to "deserialize_method_call" has incompatible type "str | None"; expected "str"  [arg-type]

Signed-off-by: a798347923 <39047817+a798347923@users.noreply.github.com>

* Update gpu_worker.py

ruff format

Signed-off-by: a798347923 <39047817+a798347923@users.noreply.github.com>

* Update core.py

ruff-format

Signed-off-by: a798347923 <39047817+a798347923@users.noreply.github.com>

* actively send exception information

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* actively send exception information

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* actively send exception information

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* change engine_core_cmd_addr(str) to engine_core_cmd_addrs(list[str]) in EngineZmqAddresses

Signed-off-by: a798347923 <2645302020@qq.com>

* change engine_core_cmd_addr(str) to engine_core_cmd_addrs(list[str]) in EngineZmqAddresses

Signed-off-by: a798347923 <2645302020@qq.com>

* Update utils.py

delete engine_core_cmd_addr in EngineZmqAddresses

Signed-off-by: a798347923 <39047817+a798347923@users.noreply.github.com>

* Remove redundant configuration: fault-pub-port

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Send pause instructions after receiving fault info in ClientGuard

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* change engine_core_guard_identities from dict[int, bytes] to list[bytes]

Signed-off-by: a798347923 <2645302020@qq.com>

* fix bug "only the worker guard of engine core 0 can receive messages sent from engine core guard

Signed-off-by: a798347923 <2645302020@qq.com>

* change local_rank to rank_in_group in WorkerGuard

Signed-off-by: a798347923 <2645302020@qq.com>

* changed del self.client_cmd_registry[int(unhealthy_engine.engine_id)]

Signed-off-by: a798347923 <2645302020@qq.com>

* add gloo communication timeout

* fix some bug

* add  stateless_process_group gloo_comm_timeout

* reconstruct fault receiver&fault handler

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* fix some bug

* reconstruct fault receiver&fault handler

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* reconstruct fault receiver&fault handler

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* fix return format

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* fix return format

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* fix return format

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* add abort request

* fix some bug

* fix some bug

* fix some bug

* add dt for client guard

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* add dt for client guard

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* add dt for client guard

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* Implementation of two types of pause: a soft one by using flag signals and a hard one by aborting nccl communicators.

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Refine certain log forms and fix a minor bug in pause function.

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Refactor and abstract the recv_msg logic in CG,ECG,WG.

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Add and check method uuid when sending commands and receiving results.

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Abstract the logic of sending instructions and waiting responses from FaultHandler

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Add options in EngineCoreGuard to recv execution results from WorkerGuard

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Support worker reinitialization after hard pause; add task queue in FaultHandler to ensure sequential task execution

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* resolve conflicts

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* resolve conflicts

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* resolve conflicts

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* resolve conflicts

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* resolve conflicts

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* resolve conflicts

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* add engine core ut

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* add engine core ut

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* Ensure WorkerGuard command execution returns result; fix missing set_device when TP>1

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* rename& format logger

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* rename& format logger

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* feat(nccl): enable non-blocking NCCL communicators to support ncclCommAbort

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* reinit dp_group

* fix bug

* fix bug

* fix bug

* fix bug (#54)

* Move requests to waiting queue instead of abandoing them directly.

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* add annotation

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* fix typos

Signed-off-by: fangyuchu <fangyuchu@qq.com>

---------

Signed-off-by: fangyuchu <fangyuchu@qq.com>
Signed-off-by: a798347923 <2645302020@qq.com>
Signed-off-by: TianZhuo <2770730562@qq.com>
Signed-off-by: a798347923 <39047817+a798347923@users.noreply.github.com>
Signed-off-by: 205150940 <112750056+205150940@users.noreply.github.com>
Signed-off-by: w00689259 <wangzhuo66@huawei.com>
Signed-off-by: zWaNg3 <37772915+zWaNg3@users.noreply.github.com>
Co-authored-by: zWaNg3 <37772915+zWaNg3@users.noreply.github.com>
Co-authored-by: a798347923 <2645302020@qq.com>
Co-authored-by: TianZhuo <2770730562@qq.com>
Co-authored-by: 205150940 <112750056+205150940@users.noreply.github.com>
Co-authored-by: a798347923 <39047817+a798347923@users.noreply.github.com>
Co-authored-by: w00689259 <wangzhuo66@huawei.com>

* Fix DT and zmq socket closing issues, updated names per feedback and reinitialize dp_group with new port

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Improve documentation and logging in API server

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Fix hanging issue in DT; fix hang when aborting communicators from Python side; use queue.Queue for engine_exception_q

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Refactor fault tolerance modules by renaming classes to Sentinel and converting engine_registry to a dict

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* reject requests when engine is in fault status

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* clear batch_queue for async scheduling

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Fix incorrect initialization of worker_cmd_socket in multi-node setups

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Switch from field to Field

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Unify start_engine_core_monitor in MPClient and CoreEngineProcManager to reduce duplication

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* refactor(Sentinel): Abstract and refactor class to standardize fault … (#84)

* refactor(Sentinel): Abstract and refactor class to standardize fault tolerance logic

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* refactor(Sentinel): Abstract and refactor class to standardize fault tolerance logic

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* refactor(Sentinel): Abstract and refactor class to standardize fault tolerance logic

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* refactor(Sentinel): Abstract and refactor class to standardize fault tolerance logic

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* refactor(Sentinel): Abstract and refactor class to standardize fault tolerance logic

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* refactor(Sentinel): Abstract and refactor class to standardize fault tolerance logic

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* refactor(Sentinel): Abstract and refactor class to standardize fault tolerance logic

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* refactor(Sentinel): Abstract and refactor class to standardize fault tolerance logic

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* refactor(Sentinel): Abstract and refactor class to standardize fault tolerance logic

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* refactor(Sentinel): Abstract and refactor class to standardize fault tolerance logic

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* refactor(Sentinel): Abstract and refactor class to standardize fault tolerance logic

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

* refactor(Sentinel): Abstract and refactor class to standardize fault tolerance logic

Signed-off-by: w00689259 <wangzhuo66@huawei.com>

---------

Signed-off-by: w00689259 <wangzhuo66@huawei.com>
Co-authored-by: w00689259 <wangzhuo66@huawei.com>
Signed-off-by: zWaNg3 <389750525@qq.com>

* fix bug in tests

Signed-off-by: w00689259 <wangzhuo66@huawei.com>
Signed-off-by: zWaNg3 <389750525@qq.com>

* fix bug in tests

Signed-off-by: w00689259 <wangzhuo66@huawei.com>
Signed-off-by: zWaNg3 <389750525@qq.com>

* refactor: improve naming and add comments for readability

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Pass fault_tolerance_config through process group creation for future extensibility

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Switch to native preempt_request implementation

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Rename base_sentinel.py

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* refactor(api_server): Split fault_tolerance interfaces into standalone files

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Use zmq poll for socket receive in Sentinel DT to avoid hanging

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Add shutdown-on-fault-tolerance-failure config option

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* ClientSentinel: add extra check to prevent repeated pause commands on error

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* feat(pause): apply pause with target index

Signed-off-by: zWaNg3 <389750525@qq.com>

* Add middleware for fault tolerance

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* fix engine_actor monitoring function bug

Signed-off-by: TianZhuo <2770730562@qq.com>

* fix engine_actor monitoring bug

Signed-off-by: TianZhuo <2770730562@qq.com>

* logger output format

Signed-off-by: TianZhuo <2770730562@qq.com>

* refactor(client_sentinel): support ClientSentinel-Client communication; refactor internal socket logic

Signed-off-by: zWaNg3 <389750525@qq.com>

* feat: add FaultToleranceRequest and FaultToleranceResult

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* feat: add EngineStatusType enum and support paused state

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Unify the logic of engine monitor for engine process manager and engine actor manager

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Fix the hanging issue of ClientSentinel in the shutdown

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* refactor(client_sentinel): rename process_ft_requests_loop function and run function

Signed-off-by: zWaNg3 <389750525@qq.com>

* Use VllmConfig as the input of Sentinel Modules

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Remove redundant @dataclass from FaultToleranceConfig

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Move hardcoded vllm_fault topic string into FaultToleranceConfig

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Update corresponding tests to new ClientSentinel design.

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Update engine core sentinel tests.

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Fix incorrect device settings in the pause of worker sentinel.

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Code cleanup and readability improvements

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Simplify FaultInfo and improve the readability

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Move sentinels into one file

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Remove recv_router_dealer_message

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Simplify the code in BaseSentinel

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Simplify the code in EngineCoreSentinel

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Introduce fault_tolerance utils and address dataclass

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* refactor: split different sentinels into separate files

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* refactor: split worker sentinel into v1/worker/sentinel for better plugin support and hardware adaptation

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* remove ThreadSafeDict

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Simplify the communication between client, client sentinel and engine core sentinel (#137)

* refactor(client_sentinel): use core_client input_socket to broadcast ft_requst

Signed-off-by: zWaNg3 <389750525@qq.com>

* refactor(client_sentinel): use core_client input_socket to broadcast ft_requst

Signed-off-by: zWaNg3 <389750525@qq.com>

* refactor(client_sentinel): use core_client input_socket to broadcast ft_requst

Signed-off-by: zWaNg3 <389750525@qq.com>

* add _send_utility_result in ClientSentinel

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Processes fault-tolerant requests and forwards them to output.

Signed-off-by: yzchang-plus <1078477584@qq.com>

* replace uncertain code with TODO

Signed-off-by: yzchang-plus <1078477584@qq.com>

* add monitoring logic in client sentinel and implement thread-safe pause in monitoring.

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* refactor(client_sentinel): send ft request using input_address

Signed-off-by: zWaNg3 <389750525@qq.com>

* refactor(client_sentinel): return ft result to client

Signed-off-by: zWaNg3 <389750525@qq.com>

* Use call_utility_async for interactions between client, client_sentinel and engine core sentinel

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Rename engine recovery timeout config

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Remove upstream and downstream concept from the base sentinel

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Support passing stateless dp port to retry

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Improve the shutdown of client sentinel

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Add try except for handle_fault in engine core

Signed-off-by: fangyuchu <fangyuchu@qq.com>

---------

Signed-off-by: zWaNg3 <389750525@qq.com>
Signed-off-by: fangyuchu <fangyuchu@qq.com>
Signed-off-by: yzchang-plus <1078477584@qq.com>
Co-authored-by: zWaNg3 <389750525@qq.com>
Co-authored-by: yzchang-plus <1078477584@qq.com>

---------

Signed-off-by: fangyuchu <fangyuchu@qq.com>
Signed-off-by: a798347923 <2645302020@qq.com>
Signed-off-by: TianZhuo <2770730562@qq.com>
Signed-off-by: a798347923 <39047817+a798347923@users.noreply.github.com>
Signed-off-by: 205150940 <112750056+205150940@users.noreply.github.com>
Signed-off-by: w00689259 <wangzhuo66@huawei.com>
Signed-off-by: zWaNg3 <37772915+zWaNg3@users.noreply.github.com>
Signed-off-by: zWaNg3 <389750525@qq.com>
Signed-off-by: yzchang-plus <1078477584@qq.com>
Co-authored-by: zWaNg3 <37772915+zWaNg3@users.noreply.github.com>
Co-authored-by: a798347923 <2645302020@qq.com>
Co-authored-by: TianZhuo <2770730562@qq.com>
Co-authored-by: 205150940 <112750056+205150940@users.noreply.github.com>
Co-authored-by: a798347923 <39047817+a798347923@users.noreply.github.com>
Co-authored-by: w00689259 <wangzhuo66@huawei.com>
Co-authored-by: zWaNg3 <389750525@qq.com>
Co-authored-by: yzchang-plus <1078477584@qq.com>
Signed-off-by: fangyuchu <fangyuchu@qq.com>

zWaNg3 committed:
refactor(dt tests of sentinels): add dt tests for sentinels

Signed-off-by: zWaNg3 <389750525@qq.com>
Signed-off-by: fangyuchu <fangyuchu@qq.com>

fangyuchu committed:
Remove torch.cuda API call (#148)

* Remove torch.cuda API call

Signed-off-by: fangyuchu <fangyuchu@qq.com>

* Remove unwanted shutdown

Signed-off-by: fangyuchu <fangyuchu@qq.com>

---------

Signed-off-by: fangyuchu <fangyuchu@qq.com>

SageMoore commented on vllm/config/parallel.py:
```diff
@@ -544,22 +554,26 @@ def stateless_init_dp_group(
                     self.data_parallel_size,
                     backend="gloo",
                     return_store=return_store,
+                    fault_tolerance_config=fault_tolerance_config,
```
Nit: can you just pass the gloo timeout instead of the whole config?

fangyuchu replied:
Revised as suggested: https://github.com/vllm-project/vllm/pull/38534/changes/2a071686d35f3ed4dfec2b2b2298e04c9f5d23ab

SageMoore reviewed (commented):
Hi @fangyuchu. Great work on this. I have an initial round of comments covering the config and UX parts of the code.

fangyuchu commented on vllm/config/fault_tolerance.py:
```diff
@@ -0,0 +1,76 @@
+    The timeout (in seconds) for gloo communication.
+    """
+
+    worker_cmd_addr: str | None = None
```
I’ve updated this to pass the address via the executor’s `collective_rpc` and create the worker sentinel on the worker side instead. 

Change is here: https://github.com/vllm-project/vllm/pull/38534/changes/527162a2d9352ba7018bc9fdbe7ae4f6083459c3

fangyuchu commented on vllm/config/fault_tolerance.py:
```diff
@@ -0,0 +1,76 @@
+    Port used to publish engine fault and status change notifications.
+    """
+
+    gloo_comm_timeout: int = 30
```
Revised as suggested.
